File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/ResourceProviderState.java
package org.maplibre.nativeffi;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import org.maplibre.nativeffi.internal.ResourceStructs;
import org.maplibre.nativeffi.internal.c.mln_resource_provider;
import org.maplibre.nativeffi.internal.c.mln_resource_provider_callback;

/** Owns runtime-scoped resource provider callback state. */
final class ResourceProviderState implements AutoCloseable {
  static final int UNKNOWN_DECISION = -1;

  private final Arena arena;
  private final ResourceProviderCallback callback;
  private final MemorySegment stub;
  private final MemorySegment descriptor;

  ResourceProviderState(ResourceProviderCallback callback) {
    this.arena = Arena.ofShared();
    this.callback = callback;
    this.stub = mln_resource_provider_callback.allocate(this::invoke, arena);
    this.descriptor = mln_resource_provider.allocate(arena);
    mln_resource_provider.size(descriptor, (int) mln_resource_provider.sizeof());
    mln_resource_provider.callback(descriptor, stub);
    mln_resource_provider.user_data(descriptor, MemorySegment.NULL);
  }

  MemorySegment descriptor() {
    return descriptor;
  }

  private int invoke(MemorySegment userData, MemorySegment request, MemorySegment handle) {
    ResourceRequestHandle requestHandle = null;
    try {
      requestHandle = new ResourceRequestHandle(handle);
      var decision = callback.handle(ResourceStructs.resourceRequest(request), requestHandle);
      return requestHandle.finishProviderDecision(decision);
    } catch (Throwable ignored) {
      if (requestHandle != null) {
        return requestHandle.finishProviderException();
      }
      return UNKNOWN_DECISION;
    }
  }

  @Override
  public void close() {
    arena.close();
  }
}

File: bindings/kotlin/src/jvmMain/kotlin/org/maplibre/nativeffi/internal/callback/ResourceProviderState.kt
package org.maplibre.nativeffi.internal.callback

import java.lang.foreign.Arena
import java.lang.foreign.Linker
import java.lang.foreign.MemorySegment
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType
import org.maplibre.nativeffi.internal.c.mln_resource_provider
import org.maplibre.nativeffi.internal.c.mln_resource_provider_callback
import org.maplibre.nativeffi.internal.loader.NativeAccess
import org.maplibre.nativeffi.resource.ResourceProviderCallback
import org.maplibre.nativeffi.resource.ResourceRequestHandle

/** Owns runtime-scoped JVM FFM resource provider callback state. */
internal class ResourceProviderState(private val callback: ResourceProviderCallback) :
  AutoCloseable {
  private val arena = Arena.ofShared()
  private val gate = CallbackGate("resource provider callbacks") { arena.close() }
  private val stub: MemorySegment
  private val descriptor: MemorySegment

  init {
    val method =
      MethodHandles.lookup()
        .findVirtual(
          ResourceProviderState::class.java,
          "invoke",
          MethodType.methodType(
            Int::class.javaPrimitiveType,
            MemorySegment::class.java,
            MemorySegment::class.java,
            MemorySegment::class.java,
          ),
        )
        .bindTo(this)
    stub = Linker.nativeLinker().upcallStub(method, callbackDescriptor, arena)
    descriptor = mln_resource_provider.allocate(arena)
    mln_resource_provider.size(descriptor, mln_resource_provider.sizeof().toInt())
    mln_resource_provider.callback(descriptor, stub)
    mln_resource_provider.user_data(descriptor, MemorySegment.NULL)
  }

  fun descriptor(): MemorySegment = descriptor

  @Suppress("UNUSED_PARAMETER")
  fun invoke(userData: MemorySegment, request: MemorySegment, handle: MemorySegment): Int {
    if (request == MemorySegment.NULL || handle == MemorySegment.NULL) return UNKNOWN_DECISION
    val lease = gate.enter() ?: return UNKNOWN_DECISION
    var requestHandle: ResourceRequestHandle? = null
    return try {
      requestHandle = ResourceRequestHandle(handle)
      val decision = callback.handle(NativeAccess.resourceRequest(request), requestHandle)
      requestHandle.finishProviderDecision(decision)
    } catch (_: Throwable) {
      requestHandle?.finishProviderException() ?: UNKNOWN_DECISION
    } finally {
      lease.close()
    }
  }

  fun checkCanClose() = gate.checkCanClose()

  override fun close() = gate.close()

  private companion object {
    private const val UNKNOWN_DECISION: Int = -1

    private val callbackDescriptor = mln_resource_provider_callback.descriptor()
  }
}
